opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDataType.java

```diff
@@ -10,26 +10,71 @@
 import java.io.Serializable;
 import java.util.HashMap;
 import java.util.LinkedHashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.function.BiConsumer;
-import lombok.EqualsAndHashCode;
 import lombok.Getter;
 import org.apache.commons.lang3.EnumUtils;
 import org.opensearch.sql.data.type.ExprCoreType;
 import org.opensearch.sql.data.type.ExprType;
+import org.opensearch.sql.data.type.WideningTypeRule;
 
 /**
  * The extension of ExprType in OpenSearch.
  */
-@EqualsAndHashCode
 public class OpenSearchDataType implements ExprType, Serializable {
 
+  /**
+   * Redefine comparison operation: class (or derived) could be compared with ExprCoreType too.
+   * Used in {@link WideningTypeRule#distance(ExprType, ExprType)}.
+   */
+  public boolean equals(final Object o) {
+    if (o == this) {
+      return true;
+    }
+    if (o instanceof ExprCoreType) {
+      return exprCoreType.equals(o);
+    }
+    if (!(o instanceof OpenSearchDataType)) {
+      return false;
+    }
+    OpenSearchDataType other = (OpenSearchDataType) o;
+    if (mappingType != null && other.mappingType != null) {
+      return mappingType.equals(other.mappingType) && exprCoreType.equals(other.exprCoreType);
+    }
+    return exprCoreType.equals(other.exprCoreType);
+  }
+
+  public int hashCode() {
+    return 42 + exprCoreType.hashCode();
```

Comment:
Would this be considered a magic number that should be defined as a constant?

Reply:
This is needed to make `OpenSearchExprValueFactory::typeActionMap` work properly. Without that override it always falls to https://github.com/Bit-Quill/opensearch-project-sql/blob/5232ad241911135a7e5b852d2b11eb70f0d5fb83/opensearch/src/main/java/org/opensearch/sql/opensearch/data/value/OpenSearchExprValueFactory.java#L208-L210

This could be simplified to always return 0 (or any other constant) to enforce `equals` check always.